### ISSUE TITLE
[dotnet-trace] Getting very few managed stacks

### ISSUE THREAD

**Issue by kouvel** [dotnet/coreclr#25015]
I’m not seeing the CPUStacks item but I see AnyStacks. However, I’m seeing very few samples in the stack view and not seeing some managed functions. This is on a trace taken for ~40 seconds:

Name | Inc % | Inc | Exc % | Exc | Fold | When | First | Last
-- | -- | -- | -- | -- | -- | -- | -- | --
test![QuickJitted]PerfViewTierEventsTest+<>c.<Main>b__0_0() | 13.4 | 9 | 0.0 | 0 | 0 | `_______________3________________` | 931.118 | 932.351
\+   test![QuickJitted]PerfViewTierEventsTest.Run(bool) | 11.9 | 8 | 0.0 | 0 | 0 | `_______________2________________` | 931.872 | 932.351
\+ Event   Microsoft-Windows-DotNETRuntime/Method/JittingStarted | 1.5 | 1 | 1.5 | 1 | 0 | `_______________0________________` | 931.118 | 931.118

There should be two of the Run(bool) methods under the top frame, one that is QuickJitted and one that is OptimizedTier1, and in total only 9 samples.

In the events view I’m seeing a lot more samples under SampleProfiler/Thread/Sample:

Event Name | Time MSec | Process Name | Rest
-- | -- | -- | --
Microsoft-DotNETCore-SampleProfiler/Thread/Sample | 2.988 | Process(6692) (6692) | ThreadID="6,444"   ProcessorNumber="0" Type="External"
Microsoft-DotNETCore-SampleProfiler/Thread/Sample | 5.044 | Process(6692) (6692) | ThreadID="6,444"   ProcessorNumber="0" Type="External"
Microsoft-DotNETCore-SampleProfiler/Thread/Sample | 6.909 | Process(6692) (6692) | ThreadID="6,444"   ProcessorNumber="0" Type="External"
Microsoft-DotNETCore-SampleProfiler/Thread/Sample | 8.808 | Process(6692) (6692) | ThreadID="6,444"   ProcessorNumber="0" Type="External"
Microsoft-DotNETCore-SampleProfiler/Thread/Sample | 10.875 | Process(6692) (6692) | ThreadID="6,444"   ProcessorNumber="0" Type="External"

...

Event Name | Time MSec | Process Name | Rest
-- | -- | -- | --
Microsoft-DotNETCore-SampleProfiler/Thread/Sample | 39,330.038 | Process(6692) (6692) | ThreadID="6,444" ProcessorNumber="0"   Type="External"
Microsoft-DotNETCore-SampleProfiler/Thread/Sample | 39,330.053 | Process(6692) (6692) | ThreadID="6,444"   ProcessorNumber="0" Type="External"

Most of them are with Type=”External”, only two are with Type=”Managed” (not sure what the type means):

Event Name | Time MSec | Process Name | Rest
-- | -- | -- | --
Microsoft-DotNETCore-SampleProfiler/Thread/Sample | 932.069 | Process(6692) (6692) | ThreadID="6,444"   ProcessorNumber="0" Type="Managed"
Microsoft-DotNETCore-SampleProfiler/Thread/Sample | 1,976.172 | Process(6692) (6692) | ThreadID="6,444"   ProcessorNumber="0" Type="Managed"

Based on the first/last times of the frames in the stack view above, all of the ones that showed up in the stack view appear to be close to the first Type=”Managed“ sample event above, not sure if there is a relation. And since the Run(bool) method had not been tiered-up by that time, that’s probably why it didn’t show up in the stack view. Wondering why there are so few samples with managed stacks in a 40 s trace.

**Comment by kouvel** [comment]
Here's the test code:

```c#
    private static void Main()
    {
        Console.WriteLine($"pid: {Process.GetCurrentProcess().Id}");
        Console.Write("Enter to start: ");
        Console.ReadLine();

        Thread t;
        ThreadStart runForever = () => Run(true);

        // Run tier 0 version of Run() continuously
        t = new Thread(runForever);
        t.IsBackground = true;
        t.Start();
        s_runStarted.WaitOne();

        // Tier-up Run()
        Run(false);
        Thread.Sleep(500);
        for (int i = 1; i < 100; ++i)
            Run(false);
        Thread.Sleep(500);

        // Run tier 1 version of Run() continuously
        t = new Thread(runForever);
        t.IsBackground = true;
        t.Start();
        s_runStarted.WaitOne();
        Thread.Sleep(500);

        // Should have reached steady state perf by now
        Thread.Sleep(2000);
        Console.Write("Enter to stop: ");
        Console.ReadLine();
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static void Run(bool forever)
    {
        s_runStarted.Set();
        do
        {
            ++s_value;
        } while (forever);
    }
```

It may so be happening that any time the runtime is suspended, the tier 0 version of the code happens to be in a static variable helper in coreclr. But even so, a stack walk should show that the tier 0 code is calling some unmanaged code in the inclusive samples, rather than not showing up at all, and it may have zero exclusive samples. The tier 1 code stays in managed code always, except for stuff like page faults, etc., most if not all samples on that thread should be exclusive samples in the managed code.

**Comment by tommcdon** [comment]
Possibly related to #25046?

**Comment by jorive** [comment]
Yes, it's related to https://github.com/dotnet/coreclr/issues/25046

**Comment by tommcdon** [comment]
@kouvel we believe this issue should be fixed as part of #25046.  Can you let us know if the issue is not resolved or close the issue as appropriate?

**Comment by tommcdon** [comment]
@kouvel we believe that this issue is fixed as part of #25046.  Please let us know if you are still seeing the problem.
